packages/core/src/libs/lohyphen/mutex.ts

```diff
@@ -0,0 +1,43 @@
+/*
+ * Inspired by https://github.com/DirtyHairy/async-mutex
+ */
+
+export class Mutex {
+  private _acquired = false;
+
+  get isLocked(): boolean {
+    return this._acquired;
+  }
+
+  private get isUnlocked(): boolean {
+    return !this.isLocked;
+  }
+
+  acquire() {
+    this._acquired = true;
+
+    this._dispatch();
+  }
+
+  waitForUnlock(): Promise<void> {
+    return new Promise((resolve) => {
```

Comment:
Can we use one Promise for all waiters? We'll create it for the first `waitForUnlock`  call and keep it in private field `_promise`. For all next invocations, we can return it. In such case, we don't need to create an array of callbacks.

Reply:
I thought about it, but I did not succeed.

Added [new test `waitForUnlock reblocks after acquiring`](https://github.com/igorkamyshev/farfetched/pull/359/commits/e79da3f8a209381eae1a59d293c2bdb39d97eb43) with a case that I failed to cover with this approach 🫠 if you can propose a code that will be pass this test and will not require more than one promise — I would appreciate it.